src/Npgsql/NpgsqlConnectionStringBuilder.cs

```diff
@@ -1724,6 +1744,146 @@ protected override void GetProperties(Hashtable propertyDescriptors)
+    { 
+        var service = serviceName ?? source?.Service ?? PostgresEnvironment.Service;
+        if (service is null)
+        {
+            return null;
+        }
+
+        if (source is not null)
+        {
+            source.Service = service;
+        }
+
+        return LoadConfigurationFromIniFile(PostgresEnvironment.UserServiceFile)
+            ?? LoadConfigurationFromIniFile(PostgresEnvironment.SystemServiceFile);
+
+        NpgsqlConnectionStringBuilder? LoadConfigurationFromIniFile(string? filePath)
+        {
+            if (filePath is null || !File.Exists(filePath))
+            {
+                return null;
+            }
+
+            using var fileStream = File.OpenRead(filePath);
+
+            var builder = source ?? new NpgsqlConnectionStringBuilder();
+            var serviceFound = false;
+            foreach (var kv in ReadIniStream(fileStream))
+            {
+                var field = GetField(kv.Key);
+                if (field is not null)
+                {
+                    serviceFound = true;
+                    if (!builder.Keys.Contains(field))
+                    {
+                        builder[field] = kv.Value;
+                    }
+                }
+            }
+
+            return serviceFound ? builder : null;
+
+            string? GetField(string key)
+            {
+                if (key.Length > service!.Length &&
+                    key.StartsWith(service, StringComparison.OrdinalIgnoreCase) &&
+                    key[service.Length] == ':')
+                {
+                    return Segment(key, service.Length + 1);
+                }
+                
+                return null;
+            }
+
+            static string Segment(string key, int prefixLength)
+            {
+                var indexOf = key.IndexOf(":", prefixLength, StringComparison.OrdinalIgnoreCase);
+                return indexOf < 0 ? key.Substring(prefixLength) : key.Substring(prefixLength, indexOf - prefixLength);
+            }
+        }
+    }
+
+    static IDictionary<string, string?> ReadIniStream(Stream stream)
+    {
+        var data = new Dictionary<string, string?>(StringComparer.OrdinalIgnoreCase);
+        using (var reader = new StreamReader(stream))
+        {
+            var sectionPrefix = string.Empty;
+
+            while (reader.Peek() != -1)
+            {
+                var rawLine = reader.ReadLine()!; // Since Peak didn't return -1, stream hasn't ended.
+                var line = rawLine.Trim();
+
+                // Ignore blank lines
+                if (string.IsNullOrWhiteSpace(line))
+                {
+                    continue;
+                }
+                // Ignore comments
+                if (line[0] is ';' or '#' or '/')
+                {
+                    continue;
+                }
+                // [Section:header]
+                if (line[0] == '[' && line[line.Length - 1] == ']')
```

Comment:
I think we can do a bit better: since we already know that we're interested in only one specific section, we can skip the lines until we find the start of the section and put everything in `Dictionary` until we find another section, at which point we just return that dictionary to fill connection string builder.

Reply:
I have tried a new approach, tell me if you find it more appropriate to our context here.